src/vote/vote.mongo.service.ts

```diff
@@ -0,0 +1,159 @@
+import { HttpException, HttpStatus, Injectable, Logger } from '@nestjs/common';
+import { Model } from 'mongoose';
+import { InjectModel } from '@nestjs/mongoose';
+import { Vote, VoteDocument } from './vote.schema';
+import { CreateVoteDto, VoteRawResponseDto, VoteRequestDto, VoteResponseDto } from './vote.dto';
+import { Poll, PollDocument } from '../poll/poll.schema';
+
+@Injectable()
+export class VoteMongoService {
+    private readonly logger = new Logger(VoteMongoService.name);
+
+    constructor(
+        @InjectModel(Vote.name) private voteModel: Model<VoteDocument>,
+        @InjectModel(Poll.name) private pollModel: Model<PollDocument>,
+    ) {
+        // do nothing
+    }
+
+    // TODO: on bot: check if user is part of role allow list, check reaction allow list
+    async validateVoteRequest(pollId: string, voteRequestDto: VoteRequestDto) {
+
+        this.logger.log(`Attempting to validate vote request for pollID: ${pollId}, with request body: ${voteRequestDto}`);
+
+        this.logger.debug('Attempting to retrieve provider_id of vote request provider');
+        const providerId = this.getProviderId();
+
+        this.logger.debug('Validating Poll ID');
+        const poll = await this.getPoll(pollId);
+        this.logger.debug('Poll ID is valid');
+
+        this.logger.debug('Checking if user has already voted on this poll with this provider_account');
+        const userVotes = await this.getVotes(pollId, providerId, voteRequestDto.provider_account_id);
+        this.logger.debug(`userVotes: Vote[] length: ${Object.keys(userVotes).length}`);
+        this.logger.debug(`userVotes: Vote[] length: ${userVotes}`);
```

Comment:
mindful to remove full votes logs in the future

Reply:
Ok, I'll switch it off for prod environment